### PR TITLE
docs on shell + signal handling for exec mode

### LINKED ISSUE #1557: Consul Template does not send signals properly to commands with spaces (when using the `sh -c` wrapper)
### Consul Template version

Run `consul-template -v` to show the version. If you are not
running the latest version, please upgrade before submitting an
issue.

```
root@5829543bcaea:/# ./consul-template -v
consul-template v0.28.0 (ae2bbca1)
```

### Configuration

```hcl
root@5829543bcaea:/# cat /config.hcl
log_level = "debug"
reload_signal = ""

exec {
  kill_signal = "SIGQUIT"
  reload_signal = "SIGHUP"
  kill_timeout = "5m"

  command = "/usr/sbin/nginx -c /etc/nginx/nginx.conf"
}
```

```
root@5829543bcaea:/# cat /etc/nginx/nginx.conf
daemon off;

worker_processes 8;

events {
  worker_connections  16192;
  multi_accept        on;
  use                 epoll;
}

env FASTLY_HEADER_KEY;

include /local/nginx/conf.d/*.conf;
```

### Command

```shell
root@5829543bcaea:/# ./consul-template -config /config.hcl
2022-03-29T00:03:28.189Z [INFO] consul-template v0.28.0 (ae2bbca1)
2022-03-29T00:03:28.189Z [INFO] (runner) creating new runner (dry: false, once: false)
2022-03-29T00:03:28.190Z [DEBUG] (runner) final config: {"Consul":{"Address":"","Namespace":"","Auth":{"Enabled":false,"Username":"","Password":""},"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":false,"Key":"","ServerName":"","Verify":true},"Token":"","TokenFile":"","Transport":{"CustomDialer":null,"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":11,"TLSHandshakeTimeout":10000000000}},"Dedup":{"Enabled":false,"MaxStale":2000000000,"Prefix":"consul-template/dedup/","TTL":15000000000,"BlockQueryWaitTime":60000000000},"DefaultDelims":{"Left":null,"Right":null},"Exec":{"Command":["/usr/sbin/nginx -c /etc/nginx/nginx.conf"],"Enabled":true,"Env":{"Denylist":[],"Custom":[],"Pristine":false,"Allowlist":[]},"KillSignal":3,"KillTimeout":300000000000,"ReloadSignal":1,"Splay":0,"Timeout":0},"KillSignal":2,"LogLevel":"debug","FileLog":{"LogFilePath":"","LogRotateBytes":0,"LogRotateDuration":86400000000000,"LogRotateMaxFiles":0},"MaxStale":2000000000,"PidFile":"","ReloadSignal":0,"Syslog":{"Enabled":false,"Facility":"LOCAL0","Name":"consul-template"},"Templates":[],"TemplateErrFatal":null,"Vault":{"Address":"","Enabled":false,"Namespace":"","RenewToken":false,"Retry":{"Attempts":12,"Backoff":250000000,"MaxBackoff":60000000000,"Enabled":true},"SSL":{"CaCert":"","CaPath":"","Cert":"","Enabled":true,"Key":"","ServerName":"","Verify":true},"Transport":{"CustomDialer":null,"DialKeepAlive":30000000000,"DialTimeout":30000000000,"DisableKeepAlives":false,"IdleConnTimeout":90000000000,"MaxIdleConns":100,"MaxIdleConnsPerHost":11,"TLSHandshakeTimeout":10000000000},"UnwrapToken":false,"DefaultLeaseDuration":300000000000},"Wait":{"Enabled":false,"Min":0,"Max":0},"Once":false,"BlockQueryWaitTime":60000000000}
2022-03-29T00:03:28.190Z [INFO] (runner) creating watcher
2022-03-29T00:03:28.191Z [INFO] (runner) starting
2022-03-29T00:03:28.191Z [DEBUG] (runner) running initial templates
2022-03-29T00:03:28.191Z [DEBUG] (runner) initiating run
2022-03-29T00:03:28.191Z [DEBUG] (runner) diffing and updating dependencies
2022-03-29T00:03:28.191Z [DEBUG] (runner) watching 0 dependencies
2022-03-29T00:03:28.191Z [DEBUG] (runner) all templates rendered
2022-03-29T00:03:28.191Z [INFO] (child) spawning: sh -c /usr/sbin/nginx -c /etc/nginx/nginx.conf
```

### Steps to reproduce

After consul-template is started as above

```shell
root@5829543bcaea:/# ps ax
  PID TTY      STAT   TIME COMMAND
    1 pts/0    Ss     0:00 /bin/bash
  636 pts/2    Ss     0:00 /bin/bash
 1669 pts/0    Sl+    0:00 ./consul-template -config /config.hcl
 1677 pts/0    S      0:00 sh -c /usr/sbin/nginx -c /etc/nginx/nginx.conf
 1678 pts/0    S      0:00 nginx: master process /usr/sbin/nginx -c /etc/nginx/nginx.conf
 1679 pts/0    S      0:00 nginx: worker process
 1680 pts/0    S      0:00 nginx: worker process
 1681 pts/0    S      0:00 nginx: worker process
 1682 pts/0    S      0:00 nginx: worker process
 1683 pts/0    S      0:00 nginx: worker process
 1684 pts/0    S      0:00 nginx: worker process
 1685 pts/0    S      0:00 nginx: worker process
 1686 pts/0    S      0:00 nginx: worker process
 1687 pts/2    R+     0:00 ps ax
```

Sent HUP to the consul-template process which should be sent down to the nginx process:
```
root@5829543bcaea:/# kill -HUP 1669
```

```shell
2022-03-29T00:05:32.480Z [DEBUG] (cli) receiving signal "hangup"
2022-03-29T00:05:32.480Z [INFO] (child) receiving signal "hangup"
2022-03-29T00:05:32.480Z [DEBUG] (cli) receiving signal "urgent I/O condition"
2022-03-29T00:05:32.483Z [INFO] (runner) child process died
2022-03-29T00:05:32.483Z [ERR] (cli) child process died with exit code -1
```

And as you can see, nginx is still running even thought consul-template said above the it "died".
```
root@5829543bcaea:/# ps ax
  PID TTY      STAT   TIME COMMAND
    1 pts/0    Ss+    0:00 /bin/bash
  636 pts/2    Ss     0:00 /bin/bash
 1678 pts/0    S      0:00 nginx: master process /usr/sbin/nginx -c /etc/nginx/nginx.conf
 1688 pts/0    S      0:00 nginx: worker process
 1689 pts/0    S      0:00 nginx: worker process
 1690 pts/0    S      0:00 nginx: worker process
 1691 pts/0    S      0:00 nginx: worker process
 1692 pts/0    S      0:00 nginx: worker process
 1693 pts/0    S      0:00 nginx: worker process
 1694 pts/0    S      0:00 nginx: worker process
 1695 pts/0    S      0:00 nginx: worker process
 1696 pts/2    R+     0:00 ps ax
```

### Expected behavior

I expected for nginx to get the SIGHUP and for consul-template to continue running and managing the nginx master process.

### Actual behavior

Consul template exists with an error `[ERR] (cli) child process died with exit code -1`

With version 0.28.0, we confirmed that switching to the array command format fixes this problem because it does not use the `sh -c` wrapper anymore. Similarly, if the exec command is just `nginx` (one word), it works as well because again it does not trigger the use case to use `sh -c`.

### PR DESCRIPTION
**Pull request by eikenb**
Fixes #1557

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -276,6 +276,25 @@ Using this method you can run as many shell commands as you need with whatever
 logic you need. Though it is suggested that if it gets too long you might want
 to wrap it in a shell script, deploy and run that.
 
+#### Shell Commands and Exec Mode
+
+Using the system shell based command has one additional caveat when used for
+the Exec mode process (the managed, executed process to which it will propagate
+signals). That is to get signals to work correctly means not only does anything
+the shell runs need to handle signals, but the shell itself needs to. This
+needs to managed by you as shells will exit upon receiving most signals.
```
Comment (mkam) [inline README.md:285]: nit: "to be managed"

File: README.md [hunk 2]
```diff
@@ -276,6 +276,25 @@ Using this method you can run as many shell commands as you need with whatever
...
+of the underlying process and to be ignored by the shell process. To get this
+you need to trap the signal in the shell with no action. For example if you had
+an underlying nginx process you wanted to run with a shell command and have the
+shell ignore it you'd do..
+
+`command = "trap '' HUP; /usr/sbin/nginx -c /etc/nginx/nginx.conf"`
+
+The `trap '' HUP;` bit is enough to get the shell to ignore the HUP signal. See
+your shell's documentation on `trap` for more details on this.
+
```
Comment (mkam) [inline README.md:297]: This example really helped me here! Just to make sure I'm understanding correctly, the idea is that if you didn't add the `trap '' HUP;`, a SIGHUP signal would have likely stopped the shell process running nginx when what we want is just nginx to process the SIGHUP.

**Review by mkam — APPROVED** [review]
LGTM

File: README.md (hunk 2 above)
Comment (mkam) [reply README.md:298]: Maybe if having the non-working version for as a comparison would help? Like saying:

`command = "/usr/sbin/nginx -c /etc/nginx/nginx.conf"`
When receiving a SIGHUP, the shell process would stop, which also stops the nginx process.

`command = "trap '' HUP; /usr/sbin/nginx -c /etc/nginx/nginx.conf"`
When receiving a SIGHUP, the shell will ignore the signal and instead nginx will process the signal and reload.

(I'm not confident on my terminology here!)

Comment (eikenb) [reply README.md:298]: Thought of another way I wanted to document as a way of dealing with this.. using `exec` to have the sub-process replace the shell process. I'm going to add this as well as add a bit more on what happens if you don't do it.

Comment (mkam) [reply README.md:315]: Looked over your latest changes and they make sense to me!

**Comment by kirooshu** [comment]
Btw, if a simple `exec` prefix solves this problem with the signals, why not do that automatically? I understand backwards compatibility is a big factor but in general I assumed it is doing some `exec` behind the scenes so that should be perfectly fine default behavior. Just a thought.

**Comment by eikenb** [comment]
Hey @kirooshu, thanks for following up.

Traditionally the command fields supported (some) shell syntax. Things like conditionals (if/case/etc) or `&&`/`||` operators. You can't automatically add an `exec` because using it with these shell syntax elements would cause an error. Eg. `if true; then echo "hi"; fi` runs now, but adding `exec` to the front of that and you get an error. If I tried to detect whether I could or not... then I'm back where we started with trying to parse the shell syntax and that is where the pain is.

Make sense?

**Pull request merged**